[wasm] Correctly set PATH for node, and python

radical opened the pull request:
This showed up on windows, if you didn't have `python` in PATH. Set this
correctly, by using the relevant properties from the packages, and
setting fallbacks in case of EMSDK_PATH being used.

Fixes https://github.com/dotnet/runtime/issues/51535 .

radical committed:
[wasm] Correctly set PATH for node, and python

This showed up on windows, if you didn't have `python` in PATH. Set this
correctly, by using the relevant properties from the packages, and
setting fallbacks in case of EMSDK_PATH being used.

radical commented:
This depends on the updated packages after https://github.com/dotnet/emsdk/pull/10 is merged.

radical committed:
[wasm] Add upstream/emscripten to PATH too, needed for emcc, for example

radical committed:
[wasm] cleanup

lewing commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -256,8 +258,10 @@
       <_PythonToolsBasePath>$(EmscriptenSdkToolsPath)python</_PythonToolsBasePath>
       <_PythonToolsVersionedPath Condition="Exists($(_PythonToolsBasePath))">$([System.IO.Directory]::GetDirectories($(_PythonToolsBasePath)))</_PythonToolsVersionedPath>
       <EmscriptenPythonToolsPath Condition="'$(_PythonToolsVersionedPath)' != ''">$(_PythonToolsVersionedPath)</EmscriptenPythonToolsPath>
+      <EmscriptenPythonBinPath   Condition="'$(EmscriptenPythonToolsPath)' != ''">$(EmscriptenPythonToolsPath)\bin\</EmscriptenPythonBinPath>
```
Is this correct?

radical replied:
It is, when using emsdk checkout, like `src/mono/wasm/emsdk/python/3.7.4-2_64bit/bin`

lewing replied:
on windows too?

radekdoulik replied:
on windows I have `src\mono\wasm\emsdk\python\3.7.4-pywin32_64bit\python.exe`

radical committed:
[wasm] fix path for `python` on windows

On windows, emsdk has python binaries in the root python folder itself, like `emsdk/python/$version/python*`, but on macOS, it is under `emsdk/python/$version/bin/python*`.

Thanks to lewing for the catch!

lewing reviewed (approved):
Looks ok to me, have we tested it on windows with the new packages?  I'd like to be very sure

lewing commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -273,7 +278,9 @@
+      <EmscriptenNodeBinPath     Condition="'$(EmscriptenNodeBinPath)' != ''"    >$([MSBuild]::NormalizeDirectory($(EmscriptenNodeBinPath)))</EmscriptenNodeBinPath>
       <EmscriptenPythonToolsPath Condition="'$(EmscriptenPythonToolsPath)' != ''">$([MSBuild]::NormalizeDirectory($(EmscriptenPythonToolsPath)))</EmscriptenPythonToolsPath>
+      <EmscriptenPythonBinPath   Condition="'$(EmscriptenPythonBinPath)' != ''"  >$([MSBuild]::NormalizeDirectory($(EmscriptenPythonBinPath)))</EmscriptenPythonBinPath>
       <EmSdkUpstreamBinPath      Condition="'$(EmSdkUpstreamBinPath)' != ''"     >$([MSBuild]::NormalizeDirectory($(EmSdkUpstreamBinPath)))</EmSdkUpstreamBinPath>
```
nit : why EmSdkUpstreamBinPath  vs EmscriptenSdkUpstreamBinPath ?

radekdoulik reviewed (approved):
LGTM.

Not sure how to properly test it on Windows. @lewing, I think you mentioned there's now a way to use the custom packages. Is it documented somewhere?

radical commented:
Don't merge this yet

radical committed:
[wasm] Rework the wasm targets' handling of emscripten paths

- Move the EMSDK_PATH specific stuff to `EmSdkRepo.props`
- And if `$EMSDK_PATH` is set, then use that. This should effectively
  override that emscripten targets from workloads, should be useful for
  testing (thanks lewing).

radical commented:
Corresponding `emsdk` PR https://github.com/dotnet/emsdk/pull/12 .

lewing reviewed (approved):
Looks ok, we really need some cross platform CI for the aot builds

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -226,38 +226,13 @@
-
-      <_UsingEMSDK_PATH>true</_UsingEMSDK_PATH>
-    </PropertyGroup>
+  <Import Project="$(MSBuildThisFileDirectory)EmSdkRepo.props" Condition="'$(EMSDK_PATH)' != ''" />
```
Excellent idea!

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-fix-paths

radical committed:
[wasm] new $(WasmUseEMSDK_PATH) to override workloads

This will be useful when you want to use a local emsdk, instead of whate
you get from the installed workload.

lewing committed:
Merge remote-tracking branch 'upstream/main' into wasm-fix-paths
